Break reconcilers out into distinct packages

negz opened the pull request:
<!--
Thank you for helping to improve Crossplane!

We strongly recommend you look through our contributor guide at https://git.io/fj2m9
if this is your first time opening a Crossplane pull request. You can find us in
https://slack.crossplane.io/messages/dev if you need any help contributing.
-->

### Description of your changes
<!--
Briefly describe what this pull request does. Be sure to direct your reviewers'
attention to anything that needs special consideration.

We love pull requests that resolve an open Crossplane issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

Fixes #
-->
This PR break all reconcilers out into distinct packages, rather than the resource package.

```go
// Before
import "github.com/crossplaneio/crossplane-runtime/pkg/resource"

r := resource.NewClaimSchedulingReconciler(...)
```

```go
// After
import "github.com/crossplaneio/crossplane-runtime/pkg/reconciler/claimscheduling"

r := claimscheduling.NewReconciler(...)
```

I'm hoping this will help us pass in loggers and eventers using variadic options without awkward names like WithClaimBindingReconcilerLogger. This is a huge diff but there's no functional changes - only moving things around and renaming them.

I was hoping we'd be able to do this without a breaking API change by using type aliases, but doing so would create cyclic imports, because the new reconciler packages depend on the resource package, which is where we'd need to create the type alias to the reconciler packages.

### Checklist
<!--
Please run through the below readiness checklist. The first two items are
relevant to every Crossplane pull request.
-->
I have:
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Ensured this PR contains a neat, self documenting set of commits.
- [ ] Updated any relevant [documentation], [examples], or [release notes].
- [ ] Updated the RBAC permissions in [`clusterrole.yaml`] to include any new types.

[documentation]: https://github.com/crossplaneio/crossplane/tree/master/docs
[examples]: https://github.com/crossplaneio/crossplane/tree/master/cluster/examples
[release notes]: https://github.com/crossplaneio/crossplane/tree/master/PendingReleaseNotes.md
[`clusterrole.yaml`]: https://github.com/crossplaneio/crossplane/blob/master/cluster/charts/crossplane/templates/clusterrole.yaml

negz commented:
Note: This will require an update to the Services Developer Guide (in the Crossplane repo) to point to the new packages. It will also require a new minor version release of crossplane-runtime, and some (boilerplate) changes to existing stacks.

hasheddan commented on pkg/reconciler/claimbinding/api.go:
```diff
@@ -125,18 +124,18 @@ func (a *APIManagedConnectionPropagator) PropagateConnection(ctx context.Context
-	})
+	k := strings.Join([]string{resource.AnnotationKeyPropagateToPrefix, string(cmcs.GetUID())}, resource.AnnotationDelimiter)
+	v := strings.Join([]string{cmcs.GetNamespace(), cmcs.GetName()}, resource.AnnotationDelimiter)
+	meta.AddAnnotations(mgcs, map[string]string{k: v})
```
Thanks for cleaning this up :)

negz replied:
Thanks! To be honest I think it was fine as is, but it started to feel pretty long after adding the `resource.` package name in a few places.

hasheddan commented on pkg/reconciler/claimbinding/configurator.go:
```diff
@@ -51,16 +52,16 @@ func NewObjectMetaConfigurator(_ runtime.ObjectTyper) *ObjectMetaConfigurator {
 
 // ConfigureNames configures the name and external name of the supplied managed
-// resource. The managed resource name is derived from the supplied resource
+// resource. The managed resource name is derived from d resource
```
I am thinking this is an unintentional change here.

negz replied:
It was - thanks!

hasheddan commented on pkg/resource/enqueue_handlers.go:
```diff
@@ -109,7 +111,7 @@ func addPropagated(obj runtime.Object, queue adder) {
 			continue
 		}
-		t := strings.Split(val, SlashDelimeter)
+		t := strings.Split(val, AnnotationDelimiter)
```
+1

negz commented:
@hasheddan PTAL - I've updated this to factor your `Target` reconciler implementation out into a distinct package like the other reconcilers. Two things of note:

* I updated the `ManagedConnectionPropagator` interface to take `LocalConnectionSecretOwner` as an argument, rather than `Claim` or `Target`. I'd forgotten it existed, and it was pretty much a subset of both `Target` and `Claim`.
* I added a second commit that makes the target reconciler's `ManagedConnectionPropagator` configurable, which lets us stub it out for slightly smaller and more loosely coupled unit tests.
